foundry.toml

```diff
@@ -0,0 +1,31 @@
+[profile.default]
+  fuzz = { seed = 1 }
+  optimizer = true
+  optimizer_runs = 10_000_000
+  remappings = ["@openzeppelin=lib/flexible-voting/lib/openzeppelin-contracts/"]
```

Comment:
_(arbitrary comment location to start a thread)_

The `GovernorCountingFractional` contract has assembly in a few spots, but that assembly is not listed as memory safe even though it could be. Options are:
1. Assume improvements are minor, ignore it (my guess is improvements will be minor)
2. Perform the same bytecode size/gas report benchmark steps listed in the previous comment for the fractional governor repo to determine significance, and use that to guide the path forward

Reply:
Without IR

| Contract                                       | Size (kB) | Margin (kB) |
|------------------------------------------------|-----------|-------------|
| AToken                                         | 13.783    | 10.793      |
| ATokenFlexVoting                               | 19.563    | 5.013       |
| AaveOracle                                     | 3.405     | 21.171      |
| Address                                        | 0.086     | 24.49       |
| Address.0.8.10                                 | 0.086     | 24.49       |
| Address.0.8.15                                 | 0.086     | 24.49       |
| BaseImmutableAdminUpgradeabilityProxy          | 1.607     | 22.969      |
| BaseUpgradeabilityProxy                        | 0.148     | 24.428      |
| Checkpoints.0.8.10                             | 0.086     | 24.49       |
| Checkpoints.0.8.15                             | 0.086     | 24.49       |
| Comet                                          | 27.055    | -2.479      |
| CometConfiguration                             | 0.063     | 24.513      |
| CometFlexVoting                                | 32.229    | -7.653      |
| CometMath                                      | 0.063     | 24.513      |
| CometStorage                                   | 1.053     | 23.523      |
| ConfiguratorInputTypes                         | 0.086     | 24.49       |
| ConfiguratorLogic                              | 8.799     | 15.777      |
| Counters.0.8.10                                | 0.086     | 24.49       |
| Counters.0.8.15                                | 0.086     | 24.49       |
| DataTypes                                      | 0.086     | 24.49       |
| DoubleEndedQueue.0.8.10                        | 0.086     | 24.49       |
| DoubleEndedQueue.0.8.15                        | 0.086     | 24.49       |
| ECDSA.0.8.10                                   | 0.086     | 24.49       |
| ECDSA.0.8.15                                   | 0.086     | 24.49       |
| ERC20.0.8.10                                   | 2.854     | 21.722      |
| ERC20.0.8.15                                   | 2.853     | 21.723      |
| Errors                                         | 7.124     | 17.452      |
| FixedPointMathLib                              | 0.086     | 24.49       |
| FractionalGovernor.0.8.10                      | 15.703    | 8.873       |
| FractionalGovernor.0.8.15                      | 15.749    | 8.827       |
| FractionalPool                                 | 6.469     | 18.107      |
| GPv2SafeERC20                                  | 0.086     | 24.49       |
| GovToken.0.8.10                                | 9.376     | 15.2        |
| GovToken.0.8.15                                | 9.374     | 15.202      |
| InitializableImmutableAdminUpgradeabilityProxy | 2.4       | 22.176      |
| InitializableUpgradeabilityProxy               | 1.216     | 23.36       |
| Math.0.8.10                                    | 0.086     | 24.49       |
| Math.0.8.15                                    | 0.086     | 24.49       |
| MockATokenFlexVoting                           | 19.843    | 4.733       |
| PercentageMath                                 | 0.086     | 24.49       |
| PoolConfigurator                               | 24.267    | 0.309       |
| ProposalReceiverMock.0.8.10                    | 0.334     | 24.242      |
| ProposalReceiverMock.0.8.15                    | 0.334     | 24.242      |
| ReserveConfiguration                           | 0.171     | 24.405      |
| SafeCast                                       | 0.086     | 24.49       |
| SafeCast.0.8.10                                | 0.086     | 24.49       |
| SafeCast.0.8.15                                | 0.086     | 24.49       |
| Strings.0.8.10                                 | 0.086     | 24.49       |
| Strings.0.8.15                                 | 0.086     | 24.49       |
| Timers.0.8.10                                  | 0.086     | 24.49       |
| Timers.0.8.15                                  | 0.086     | 24.49       |
| WadRayMath                                     | 0.086     | 24.49       |
| console.0.8.10                                 | 0.086     | 24.49       |
| console.0.8.15                                 | 0.086     | 24.49       |
| console.0.8.20                                 | 0.086     | 24.49       |
| console2.0.8.10                                | 0.086     | 24.49       |
| console2.0.8.15                                | 0.086     | 24.49       |
| console2.0.8.20                                | 0.086     | 24.49       |
| stdError.0.8.10                                | 0.72      | 23.856      |
| stdError.0.8.15                                | 0.72      | 23.856      |
| stdError.0.8.20                                | 0.713     | 23.863      |
| stdJson.0.8.10                                 | 0.086     | 24.49       |
| stdJson.0.8.15                                 | 0.086     | 24.49       |
| stdJson.0.8.20                                 | 0.086     | 24.49       |
| stdMath.0.8.10                                 | 0.086     | 24.49       |
| stdMath.0.8.15                                 | 0.086     | 24.49       |
| stdMath.0.8.20                                 | 0.086     | 24.49       |
| stdStorage.0.8.10                              | 0.086     | 24.49       |
| stdStorage.0.8.15                              | 0.086     | 24.49       |
| stdStorage.0.8.20                              | 0.086     | 24.49       |
| stdStorageSafe.0.8.10                          | 0.086     | 24.49       |
| stdStorageSafe.0.8.15                          | 0.086     | 24.49       |
| stdStorageSafe.0.8.20                          | 0.086     | 24.49       |

Reply:
| src/CometFlexVoting.sol:CometFlexVoting contract |                 |        |        |        |         |
|--------------------------------------------------|-----------------|--------|--------|--------|---------|
| Deployment Cost                                  | Deployment Size |        |        |        |         |
| 6572778                                          | 38710           |        |        |        |         |
| Function Name                                    | min             | avg    | median | max    | # calls |
| GOVERNOR                                         | 315             | 315    | 315    | 315    | 1       |
| balanceOf                                        | 1288            | 3900   | 2993   | 7288   | 6       |
| baseToken                                        | 293             | 293    | 293    | 293    | 1       |
| borrowBalanceOf                                  | 1588            | 3028   | 3028   | 4468   | 2       |
| castVote                                         | 962             | 44765  | 55076  | 69396  | 25      |
| expressVote                                      | 2516            | 36750  | 47845  | 48358  | 38      |
| governor                                         | 340             | 340    | 340    | 340    | 1       |
| initializeStorage                                | 44730           | 44730  | 44730  | 44730  | 34      |
| proposalVotes(uint256)(uint128,uint128,uint128)  | 2791            | 2791   | 2791   | 2791   | 2       |
| proposalVotes(uint256)(uint256,uint256,uint256)  | 791             | 2391   | 2791   | 2791   | 5       |
| supply                                           | 68563           | 136921 | 183833 | 195733 | 54      |
| withdraw                                         | 15872           | 100570 | 127915 | 127915 | 11      |


| src/FractionalPool.sol:FractionalPool contract  |                 |        |        |        |         |
|-------------------------------------------------|-----------------|--------|--------|--------|---------|
| Deployment Cost                                 | Deployment Size |        |        |        |         |
| 1311222                                         | 6819            |        |        |        |         |
| Function Name                                   | min             | avg    | median | max    | # calls |
| TOKEN                                           | 249             | 249    | 249    | 249    | 1       |
| borrow                                          | 51506           | 58545  | 53506  | 75665  | 4       |
| borrowTotal                                     | 519             | 519    | 519    | 519    | 1       |
| castVote                                        | 1532            | 53210  | 67352  | 86417  | 10      |
| deposit                                         | 61445           | 137888 | 167694 | 167694 | 20      |
| deposits                                        | 587             | 1253   | 587    | 2587   | 3       |
| expressVote                                     | 2371            | 38914  | 47845  | 48357  | 15      |
| getPastDeposits                                 | 1778            | 1778   | 1778   | 1778   | 2       |
| getPastTotalDeposits                            | 1564            | 1564   | 1564   | 1564   | 1       |
| internalVotingPeriodEnd                         | 1427            | 1427   | 1427   | 1427   | 7       |
| proposalVotes(uint256)(uint128,uint128,uint128) | 724             | 1390   | 724    | 2724   | 3       |
| proposalVotes(uint256)(uint256,uint256,uint256) | 2724            | 2724   | 2724   | 2724   | 2       |


| test/FractionalGovernor.sol:FractionalGovernor contract |                 |       |        |       |         |
|---------------------------------------------------------|-----------------|-------|--------|-------|---------|
| Deployment Cost                                         | Deployment Size |       |        |       |         |
| 3179132                                                 | 16883           |       |        |       |         |
| Function Name                                           | min             | avg   | median | max   | # calls |
| COUNTING_MODE                                           | 703             | 703   | 703    | 703   | 1       |
| EXTENDED_BALLOT_TYPEHASH                                | 307             | 307   | 307    | 307   | 2       |
| castVoteWithReasonAndParams                             | 5025            | 39310 | 41663  | 78173 | 48      |
| castVoteWithReasonAndParamsBySig                        | 6089            | 34922 | 17526  | 83226 | 5       |
| execute                                                 | 5652            | 17579 | 17630  | 29406 | 4       |
| exposed_quorumReached                                   | 1094            | 3094  | 3094   | 5094  | 8       |
| exposed_setFractionalVoteNonce                          | 22647           | 22647 | 22647  | 22647 | 2       |
| fractionalVoteNonce                                     | 633             | 633   | 633    | 633   | 9       |
| hasVoted                                                | 692             | 1692  | 1692   | 2692  | 26      |
| hashProposal                                            | 3639            | 3639  | 3639   | 3639  | 26      |
| name                                                    | 3349            | 3349  | 3349   | 3349  | 2       |
| proposalDeadline                                        | 627             | 627   | 627    | 627   | 23      |
| proposalSnapshot                                        | 608             | 608   | 608    | 608   | 61      |
| proposalVotes(uint256)(uint128,uint128,uint128)         | 706             | 1069  | 706    | 4706  | 11      |
| proposalVotes(uint256)(uint256,uint256,uint256)         | 706             | 1081  | 706    | 4706  | 16      |
| propose                                                 | 61096           | 61169 | 61096  | 63596 | 34      |
| quorum                                                  | 362             | 362   | 362    | 362   | 8       |
| state                                                   | 1436            | 2281  | 2416   | 3117  | 72      |
| token                                                   | 314             | 314   | 314    | 314   | 2       |
| voteWeightCast                                          | 678             | 1106  | 678    | 2678  | 14      |
| votingDelay                                             | 260             | 260   | 260    | 260   | 45      |
| votingPeriod                                            | 250             | 250   | 250    | 250   | 23      |


| test/FractionalGovernor.sol:FractionalGovernor.0.8.10 contract |                 |       |        |       |         |
|----------------------------------------------------------------|-----------------|-------|--------|-------|---------|
| Deployment Cost                                                | Deployment Size |       |        |       |         |
| 3167944                                                        | 16725           |       |        |       |         |
| Function Name                                                  | min             | avg   | median | max   | # calls |
| castVoteWithReasonAndParams                                    | 7291            | 52048 | 58185  | 78085 | 46      |
| proposalDeadline                                               | 627             | 627   | 627    | 627   | 6       |
| proposalSnapshot                                               | 608             | 608   | 608    | 608   | 177     |
| proposalVotes(uint256)(uint128,uint128,uint128)                | 706             | 991   | 706    | 4706  | 14      |
| proposalVotes(uint256)(uint256,uint256,uint256)                | 706             | 1119  | 706    | 4706  | 29      |
| propose                                                        | 61189           | 61333 | 61189  | 63689 | 52      |
| state                                                          | 1430            | 2270  | 2270   | 3111  | 104     |
| token                                                          | 314             | 314   | 314    | 314   | 108     |


| test/FractionalGovernor.sol:FractionalGovernor.0.8.15 contract |                 |       |        |       |         |
|----------------------------------------------------------------|-----------------|-------|--------|-------|---------|
| Deployment Cost                                                | Deployment Size |       |        |       |         |
| 3177156                                                        | 16882           |       |        |       |         |
| Function Name                                                  | min             | avg   | median | max   | # calls |
| castVoteWithReasonAndParams                                    | 7297            | 50784 | 58194  | 78094 | 25      |
| proposalDeadline                                               | 627             | 627   | 627    | 627   | 6       |
| proposalSnapshot                                               | 608             | 608   | 608    | 608   | 100     |
| proposalVotes(uint256)(uint128,uint128,uint128)                | 706             | 1277  | 706    | 4706  | 7       |
| proposalVotes(uint256)(uint256,uint256,uint256)                | 706             | 1456  | 706    | 4706  | 16      |
| propose                                                        | 61189           | 61430 | 61189  | 63689 | 31      |
| state                                                          | 1436            | 2276  | 2276   | 3117  | 62      |
| token                                                          | 314             | 314   | 314    | 314   | 56      |


| test/GovToken.sol:GovToken contract |                 |       |        |       |         |
|-------------------------------------|-----------------|-------|--------|-------|---------|
| Deployment Cost                     | Deployment Size |       |        |       |         |
| 1918563                             | 10134           |       |        |       |         |
| Function Name                       | min             | avg   | median | max   | # calls |
| approve                             | 2626            | 22530 | 24626  | 24626 | 21      |
| balanceOf                           | 562             | 962   | 562    | 2562  | 15      |
| delegate                            | 22825           | 45530 | 26825  | 69702 | 75      |
| delegates                           | 2576            | 2576  | 2576   | 2576  | 1       |
| exposed_maxSupply                   | 259             | 259   | 259    | 259   | 2       |
| exposed_mint                        | 8752            | 71324 | 96306  | 96306 | 57      |
| getPastVotes                        | 1610            | 2101  | 1610   | 2818  | 94      |
| getVotes                            | 1094            | 1094  | 1094   | 1094  | 1       |
| name                                | 3221            | 3221  | 3221   | 3221  | 1       |
| symbol                              | 3219            | 3219  | 3219   | 3219  | 1       |
| transfer                            | 28547           | 35586 | 30547  | 52706 | 4       |
| transferFrom                        | 5661            | 46206 | 59784  | 59784 | 20      |


| test/GovToken.sol:GovToken.0.8.10 contract |                 |       |        |       |         |
|--------------------------------------------|-----------------|-------|--------|-------|---------|
| Deployment Cost                            | Deployment Size |       |        |       |         |
| 1922981                                    | 10028           |       |        |       |         |
| Function Name                              | min             | avg   | median | max   | # calls |
| approve                                    | 2622            | 23724 | 24622  | 24622 | 98      |
| balanceOf                                  | 562             | 583   | 562    | 2562  | 92      |
| decimals                                   | 267             | 267   | 267    | 267   | 65      |
| delegate                                   | 26825           | 28716 | 26825  | 69697 | 68      |
| delegates                                  | 2576            | 2576  | 2576   | 2576  | 1       |
| exposed_mint                               | 26893           | 76239 | 96296  | 96296 | 98      |
| getPastVotes                               | 1592            | 2089  | 1592   | 2818  | 141     |
| transfer                                   | 5296            | 30067 | 30519  | 52686 | 28      |
| transferFrom                               | 5644            | 43971 | 59780  | 74725 | 96      |


| test/GovToken.sol:GovToken.0.8.15 contract |                 |       |        |       |         |
|--------------------------------------------|-----------------|-------|--------|-------|---------|
| Deployment Cost                            | Deployment Size |       |        |       |         |
| 1922563                                    | 10154           |       |        |       |         |
| Function Name                              | min             | avg   | median | max   | # calls |
| approve                                    | 2622            | 23302 | 24622  | 24622 | 50      |
| balanceOf                                  | 562             | 698   | 562    | 2562  | 44      |
| decimals                                   | 267             | 267   | 267    | 267   | 34      |
| delegate                                   | 26825           | 30301 | 26825  | 69697 | 37      |
| delegates                                  | 2576            | 2576  | 2576   | 2576  | 1       |
| exposed_mint                               | 26893           | 75633 | 96296  | 96296 | 50      |
| getPastVotes                               | 1592            | 2148  | 2040   | 2818  | 78      |
| transfer                                   | 757             | 24090 | 30519  | 30519 | 10      |
| transferFrom                               | 5644            | 43081 | 59780  | 72725 | 47      |


| test/MockATokenFlexVoting.sol:MockATokenFlexVoting contract |                 |        |        |        |         |
|-------------------------------------------------------------|-----------------|--------|--------|--------|---------|
| Deployment Cost                                             | Deployment Size |        |        |        |         |
| 4031769                                                     | 20827           |        |        |        |         |
| Function Name                                               | min             | avg    | median | max    | # calls |
| approve                                                     | 24565           | 24565  | 24565  | 24565  | 1       |
| balanceOf                                                   | 2623            | 3584   | 3061   | 18061  | 67      |
| burn                                                        | 15516           | 51661  | 41295  | 105658 | 10      |
| castVote                                                    | 994             | 46762  | 55083  | 69403  | 47      |
| exposed_RawBalanceOf                                        | 603             | 603    | 603    | 603    | 11      |
| exposed_Treasury                                            | 2355            | 2355   | 2355   | 2355   | 1       |
| expressVote                                                 | 2489            | 35243  | 47798  | 48342  | 73      |
| getPastRawBalance                                           | 1705            | 1837   | 1705   | 2216   | 15      |
| getPastTotalBalance                                         | 1562            | 1755   | 1562   | 2073   | 20      |
| handleRepayment                                             | 467             | 467    | 467    | 467    | 3       |
| initialize                                                  | 195580          | 195580 | 195580 | 195580 | 65      |
| mint                                                        | 56534           | 130989 | 155976 | 155976 | 93      |
| mintToTreasury                                              | 97810           | 97810  | 97810  | 97810  | 1       |
| name                                                        | 3334            | 3334   | 3334   | 3334   | 1       |
| proposalVotes(uint256)(uint128,uint128,uint128)             | 769             | 769    | 769    | 769    | 1       |
| proposalVotes(uint256)(uint256,uint256,uint256)             | 2769            | 2769   | 2769   | 2769   | 6       |
| scaledBalanceOf                                             | 626             | 626    | 626    | 626    | 14      |
| symbol                                                      | 3354            | 3354   | 3354   | 3354   | 1       |
| transfer                                                    | 63227           | 113528 | 120388 | 142606 | 10      |
| transferFrom                                                | 143164          | 143164 | 143164 | 143164 | 1       |
| transferUnderlyingTo                                        | 31483           | 34492  | 33483  | 53650  | 18      |


| test/ProposalReceiverMock.sol:ProposalReceiverMock contract |                 |      |        |      |         |
|-------------------------------------------------------------|-----------------|------|--------|------|---------|
| Deployment Cost                                             | Deployment Size |      |        |      |         |
| 65517                                                       | 359             |      |        |      |         |
| Function Name                                               | min             | avg  | median | max  | # calls |
| mockRecieverFunction                                        | 1217            | 1217 | 1217   | 1217 | 2       |

Reply:
With IR I got this error

```
Compiler run failed:
Error: Yul exception:Variable _18 is 1 too deep in the stack [ _18 expr_mpos_3 _3 _17 _1 _15 expr_component_mpos_1 _9 RET var_ins_mpos _4 _6 _2 _13 var_i expr_1 _27 ]

```

Reply:
Let's ignore this, I don't think the upside is worthwhile. It's something we might want to consider separately for a dot release of Flexible Voting.

Reply:
@alexkeating I'm a bit confusing, are these benchmarks from adding the memory safe marker, or from compiling with via-ir? They are independent

Either way, I agree with @apbendi we can ignore it for this release. I've created https://github.com/ScopeLift/flexible-voting/issues/52 to track

Reply:
Oh yea, they are for via ir. I misread bullet point 2.

Reply:
I am going to resolve then